sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,333 @@
+// Licensed under the MIT License.
+
+package com.azure.messaging.eventhubs.implementation;
+
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+import com.azure.core.implementation.util.ImplUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.messaging.eventhubs.EventHubAsyncClient;
+import com.azure.messaging.eventhubs.EventHubAsyncConsumer;
+import com.azure.messaging.eventhubs.EventProcessor;
+import com.azure.messaging.eventhubs.PartitionManager;
+import com.azure.messaging.eventhubs.models.PartitionOwnership;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+/**
+ * This class is responsible for balancing the load of processing events from all partitions of an Event Hub by
+ * distributing the number of partitions uniformly among all the  active {@link EventProcessor}s.
+ * <p>
+ * This load balancer will retrieve partition ownership details from the {@link PartitionManager} to find the number of
+ * active {@link EventProcessor}s. It uses the last modified time to decide if an EventProcessor is active. If a
+ * partition ownership entry has not be updated for a specified duration of time, the owner of that partition is
+ * considered inactive and the partition is available for other EventProcessors to own.
+ * </p>
+ * More details can be found <a href="https://gist.github.com/srnagar/7ef8566cfef0673288275c450dc99590">here</a>.
+ */
+public final class PartitionBasedLoadBalancer {
+
+    private static final Random RANDOM = new Random();
+    private final ClientLogger logger = new ClientLogger(PartitionBasedLoadBalancer.class);
+
+    private final String eventHubName;
+    private final String consumerGroupName;
+    private final PartitionManager partitionManager;
+    private final EventHubAsyncClient eventHubAsyncClient;
+    private final String ownerId;
+    private final long inactiveTimeLimitInSeconds;
+    private final PartitionPumpManager partitionPumpManager;
+
+    /**
+     * Creates an instance of PartitionBasedLoadBalancer for the given Event Hub name and consumer group.
+     *
+     * @param partitionManager The partition manager that this load balancer will use to read/update ownership details.
+     * @param eventHubAsyncClient The asynchronous Event Hub client used to consume events.
+     * @param eventHubName The Event Hub name the {@link EventProcessor} is associated with.
+     * @param consumerGroupName The consumer group name the {@link EventProcessor} is associated with.
+     * @param ownerId The owner identifier for the {@link EventProcessor} this load balancer is associated with.
```

Comment:
nit:  This reads a bit awkward.  Maybe consider "The identifier of the `Event Processor` that owns this load balancer" or similar?